tiny_dnn/core/framework/tensor.h

```diff
@@ -148,58 +150,57 @@ class Tensor {
 #endif
 
     // Returns the tensor shape
-    const std::array<size_t, 4>& shape() const { return shape_; }
+    const std::array<size_t, kDimensions>& shape() const { return shape_; }
 
     // Returns the value of a specified index in the tensor.
     // Checked version (throw exceptions for out-of-range error)
-    U& host_at(const size_t d0,
-               const size_t d1,
-               const size_t d2,
-               const size_t d3) {
-        return *host_ptr(d0, d1, d2, d3);
+    template<typename... Args>
+    U& host_at(const Args... args) {
+        return *host_ptr(args...);
+    }
+    template<typename... Args>
+    U host_at(const Args... args) const {
+        return *host_ptr(args...);
     }
 
-    U host_at(const size_t d0,
-              const size_t d1,
-              const size_t d2,
-              const size_t d3) const {
-        return *host_ptr(d0, d1, d2, d3);
+    size_t host_pos(const size_t d) {
+        if (d >= *shape_.rbegin())  {
+            throw nn_error("Access tensor out of range.");
+        }
+        return d;
     }
 
-    // Returns the pointer to a specified index in the tensor
-    // Checked version (throw exceptions for out-of-range error)
-    const U* host_ptr(const size_t d0,
-                      const size_t d1,
-                      const size_t d2,
-                      const size_t d3) const {
-        if (d0 >= shape_[0] || d1 >= shape_[1] ||
-            d2 >= shape_[2] || d3 >= shape_[3]) {
+    template<typename... Args>
+    size_t host_pos(const size_t d,
+                const Args... args) {
+        size_t dim = shape_.size() - sizeof...(args) - 1;
```

Comment:
@beru  @edgarriba then we probably should remove checks to the lower level, i.e. `host_pos`, and create separate set of non checked functions I think